ecosystem/platform/server/app/controllers/users/omniauth_callbacks_controller.rb

```diff
@@ -5,40 +5,26 @@
 
 module Users
   class OmniauthCallbacksController < Devise::OmniauthCallbacksController
-    def github
-      @user = User.from_omniauth(auth_data, current_user)
-      if @user.persisted?
-        @message = I18n.t 'devise.omniauth_callbacks.success', kind: 'Github'
-        sign_in(@user)
-      else
-        # TODO: make this bulletproof
-        @message = I18n.t 'devise.omniauth_callbacks.failure', kind: :github,
-                                                               reason: @user.errors.full_messages.join("\n")
+    User.omniauth_providers.each do |provider|
+      define_method provider do
+        oauth_callback(provider)
       end
-
-      # TODO: RENDER SOME RESPONSE!
-      # render json: { message: }
-      render 'api/users/show', formats: [:json]
     end
 
-    def discord
+    private
+
+    def oauth_callback(provider)
       @user = User.from_omniauth(auth_data, current_user)
 
       if @user.persisted?
-        message = I18n.t 'devise.omniauth_callbacks.success', kind: 'Discord'
-        sign_in(@user)
+        sign_in_and_redirect @user, event: :authentication # this will throw if @user is not activated
+        set_flash_message(:notice, :success, kind: provider.to_s.titleize) if is_navigational_format?
       else
         # TODO: make this bulletproof
-        message = I18n.t 'devise.omniauth_callbacks.failure', kind: :discord,
-                                                              reason: @user.errors.full_messages.join("\n")
+        raise "Unable to persist user"
```

Comment:
In what situation would `from_omniauth` fail to persist the user?

Reply:
when would  `event: :authentication # this will throw if @user is not activated` throw- if the user doesn't have email set yet?

Reply:
It's for warden callbacks, which since we aren't using, I removed it.